File: sdk/python/nexus_py/sessions.py
"""
nexus_py.sessions
────────────────────
SessionsClient — handles all session lifecycle endpoints.

Accessed via `nexus.sessions.*` on a NexusClient instance.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from nexus_py.models import PlayerResponse, SessionResponse

if TYPE_CHECKING:
    from nexus_py.client import NexusClient


class SessionsClient:
    """
    Session lifecycle operations: create, join, leave, lock, end, state.

    All methods that return session data return a fully-typed
    `SessionResponse` — no raw dicts ever reach the caller.
    """

    def __init__(self, client: "NexusClient") -> None:
        self._c = client

    async def create(
        self,
        game_id: str,
        config: dict[str, Any] | None = None,
    ) -> SessionResponse:
        """
        Create a new game session.

        Args:
            game_id: UUID of the registered game this session belongs to.
            config:  Game-specific session configuration (mode, max_players,
                     region, npcs, difficulty, etc).

        Returns:
            SessionResponse with a freshly generated `join_code` (format:
            "NEXUS-XXXX") and the creator registered as host.
        """
        data = await self._c._post(
            "/sessions",
            json={"game_id": game_id, "config": config or {}},
        )
        return SessionResponse(**data)

    async def get(self, session_id: str) -> SessionResponse:
        """
        Retrieve full session details by UUID.

        Raises:
            NexusError: session not found (404).
        """
        data = await self._c._get(f"/sessions/{session_id}")
        return SessionResponse(**data)

    async def join(self, session_id: str) -> SessionResponse:
        """
        Join a session by its UUID.

        Raises:
            SessionError: session full, already a member, or session ended (409).
            AuthError:    session is locked (423).
        """
        data = await self._c._post(f"/sessions/{session_id}/join")
        return SessionResponse(**data)

    async def join_by_code(self, join_code: str) -> SessionResponse:
        """
        Join a session using its human-readable join code (e.g. "NEXUS-AB23").

        The join code is case-insensitive.

        Raises:
            NexusError:   no session found with that code (404).
            SessionError: session full, already a member, or session ended (409).
            AuthError:    session is locked (423).
        """
        data = await self._c._post(f"/sessions/join/{join_code}")
        return SessionResponse(**data)

    async def leave(self, session_id: str) -> SessionResponse:
        """
        Leave a session you are currently a member of.

        Raises:
            SessionError: not currently a member of this session (409).
        """
        data = await self._c._post(f"/sessions/{session_id}/leave")
        return SessionResponse(**data)

    async def lock(self, session_id: str) -> SessionResponse:
        """
        Lock a session to prevent new players from joining.

        Only the session host may call this.

        Raises:
            NexusError: caller is not the host (403).
        """
        data = await self._c._post(f"/sessions/{session_id}/lock")
        return SessionResponse(**data)

    async def end(self, session_id: str) -> SessionResponse:
        """
        End a session permanently. Only the session host may call this.

        All active player memberships are closed. The session becomes
        immutable after this call.

        Raises:
            NexusError: caller is not the host (403).
        """
        data = await self._c._post(f"/sessions/{session_id}/end")
        return SessionResponse(**data)

    async def update_state(
        self,
        session_id: str,
        state: dict[str, Any],
    ) -> SessionResponse:
        """
        Shallow-merge `state` into the session's existing game state.

        `{ **current_state, **state }` — top-level keys in `state` overwrite
        existing keys; nested dicts are replaced wholesale, not deep-merged.

        Raises:
            SessionError: session has already ended (409).
        """
        data = await self._c._patch(
            f"/sessions/{session_id}/state",
            json={"state": state},
        )
        return SessionResponse(**data)

    async def list_players(self, session_id: str) -> list[PlayerResponse]:
        """
        List currently active players in a session.

        Note: the underlying API endpoint returns SessionPlayerResponse
        objects (which include role/joined_at/left_at), but this SDK method
        projects them down to the simpler PlayerResponse shape per the
        spec's documented return type. Use `.get(session_id).players` for
        the full membership detail including role and join timestamps.
        """
        data = await self._c._get(f"/sessions/{session_id}/players")
        return [
            PlayerResponse(
                id=p["player_id"],
                username=p["username"],
                email="",  # not present on SessionPlayerResponse — see docstring
                is_guest=False,
            )
            for p in data
        ]

File: sdk/javascript/src/sessions.ts
/**
 * nexus-js / sessions.ts
 * ────────────────────────
 * SessionsClient — session lifecycle endpoints.
 *
 * Accessed via `nexus.sessions.*` on a NexusClient instance.
 *
 * Note: the spec's reference implementation types `leave`, `lock`, and
 * `end` as `Promise<void>`, but the actual API (api/routers/sessions.py)
 * returns a full SessionResponse body (200, not 204) for all three —
 * so callers can inspect `is_locked`, `status`, `ended_at`, etc. without
 * a follow-up GET. Typed accordingly here.
 */

import type { NexusClient } from "./client";
import type { PlayerResponse, SessionResponse } from "./types";

export class SessionsClient {
  constructor(private c: NexusClient) {}

  /**
   * Create a new game session.
   *
   * Returns a SessionResponse with a freshly generated `join_code`
   * (format: "NEXUS-XXXX") and the creator registered as host.
   */
  create(gameId: string, config?: Record<string, unknown>): Promise<SessionResponse> {
    return this.c.fetch<SessionResponse>("/sessions", {
      method: "POST",
      body: JSON.stringify({ game_id: gameId, config: config ?? {} }),
    });
  }

  /** Retrieve full session details by UUID. Throws NexusError (404) if not found. */
  get(sessionId: string): Promise<SessionResponse> {
    return this.c.fetch<SessionResponse>(`/sessions/${sessionId}`);
  }

  /**
   * Join a session by its UUID.
   * Throws NexusError (409 full/already-joined/ended, or 423 locked).
   */
  join(sessionId: string): Promise<SessionResponse> {
    return this.c.fetch<SessionResponse>(`/sessions/${sessionId}/join`, { method: "POST" });
  }

  /**
   * Join a session using its human-readable join code (e.g. "NEXUS-AB23").
   * The join code is case-insensitive.
   */
  joinByCode(joinCode: string): Promise<SessionResponse> {
    return this.c.fetch<SessionResponse>(`/sessions/join/${joinCode}`, { method: "POST" });
  }

  /** Leave a session you are currently a member of. */
  leave(sessionId: string): Promise<SessionResponse> {
    return this.c.fetch<SessionResponse>(`/sessions/${sessionId}/leave`, { method: "POST" });
  }

  /** Lock a session to prevent new players from joining. Host-only. */
  lock(sessionId: string): Promise<SessionResponse> {
    return this.c.fetch<SessionResponse>(`/sessions/${sessionId}/lock`, { method: "POST" });
  }

  /** End a session permanently. Host-only. All active memberships are closed. */
  end(sessionId: string): Promise<SessionResponse> {
    return this.c.fetch<SessionResponse>(`/sessions/${sessionId}/end`, { method: "POST" });
  }

  /**
   * Shallow-merge `state` into the session's existing game state.
   *
   * `{ ...currentState, ...state }` — top-level keys in `state` overwrite
   * existing keys; nested objects are replaced wholesale, not deep-merged.
   */
  updateState(sessionId: string, state: Record<string, unknown>): Promise<SessionResponse> {
    return this.c.fetch<SessionResponse>(`/sessions/${sessionId}/state`, {
      method: "PATCH",
      body: JSON.stringify({ state }),
    });
  }

  /** List currently active players in a session. */
  listPlayers(sessionId: string): Promise<PlayerResponse[]> {
    return this.c.fetch<PlayerResponse[]>(`/sessions/${sessionId}/players`);
  }
}
